README.md
# My-PyChess
This is a **fully featured chess app** written purely in Python using Pygame Library.

![main image](screenshots/main.jpg)

Click [here](screenshots/screenshots.md) to see more few screenshots of My-PyChess in action!

Any bug-reports, suggestions or questions, you can leave it in the github issues section.
If you want to directly communicate with me, you can mail me: itsankith26@gmail.com

The My-PyChess project is available under MIT License. The MIT Licence applies to all the resources I have created in this Project. This includes all the Python files and text files. But some resources(images, sounds and font file) are not created by me, I have downloaded these from the internet. I have given credits to the authors of these resources in [this file](res/CREDITS.txt). All these resources maintain the original licenses that the authors have leased them under (These licenses permit my use of the respective resources in this project).

## Getting started, with Python source code(All OS support)
- Make sure you have **Python** and **pygame** installed and working.
- Clone this repository (or download zip file and extract it).
- Then, run the **pychess.py** file. Trying to run any other file will not run the game.
- Those who are running older versions of this app, can upgrade to the latest version by the same method.

## Getting started, by downloading executable(Windows OS only)
- If you are only interested in chess and not python programming, head to releases section on github. 
- There, you will find executable packages under the respective versions. 
- Download the My-PyChess-win-exe.zip from the latest version, extract the folder into your computer.
- Run My-PyChess.exe
- Those who are running older versions of this app, can upgrade to the latest version by the same method.

Interested in python game development with pygame but want to start off with a simpler project. I have released a [lite implementation](https://github.com/ankith26/My-PyChess-lite/) of My-PyChess, that focuses just on chess programming - free from all the code for menus, singleplayer, online etc. 

## Features
- Clean GUI with a lot of menus for ease of use.
- Allows users to make only valid move, does not allow users to make moves that puts the user's king at check.
- Detects check, checkmate, stalemate and informs user.
- Supports things like castling, pawn promotion, enpassent etc.
- Supports saving and loading games.
- Has single player mode with two different types, levels and ability to play against the stockfish chess engine.
- Has online gamemode, play chess with anyone in the world.
- Has a chess game timer.
- Has a preference menu where you can customize the app to meet your preferences.
- Has a chess howto, about menu and stockfish install/configure menu to make things easy for users.

## What's new in this Version 3.2
- Added a Back-Button, to go back to the previous menu. In older versions, the quit button was used for this purpose, but from now on a dedicated button to go back is there in the top-right corner. The quit button will now be used only to exit the app.
- Added a game timer to multiplayer mode, with a new menu to setup the timer.
- Fixed bugs and made many additions and improvements to the client-server in chess online gamemode.
- Made optimisations to core chess module and gui module.
- Added a chess hotwto.
- Minor improvements to game sounds and textbox.
- Upgraded preference menu and made the loadgame interface more robust.
- Several other minor changes and improvements made.

## Highlights of v3.x, the latest major release.
- The code was revamped and restructured, fixed bugs and made MAJOR PERFORMANCE IMPROVEMENTS.
- Online play was added. This features a ONLINE LOBBY and support upto 10 people to play chess.
- Singleplayer saw BIG UPGRADES: Firstly, a decent PYTHON CHESS ENGINE that playes chess was added.
- You can play against STOCKFISH CHESS ENGINE (see https://stockfishchess.org). My-PyChess will act as an interface to give you this singleplayer mode.
- EnPassant in chess, undo move, screenflip and sounds were added.
- LOTS of changes made to GUI.

Click [here](CHANGELOG.md) to see full changelog.

## Online Gameplay
- You can self-host the My-PyChess online server(read more [here](onlinehowto.txt) ).
- Apart from that, I have launched a public My-PyChess server for PUBLIC BETA TESTING, which ANYONE in the world can connect to.

- Caveat: The server is on a IPv6-only network. This means that your network MUST SUPPORT IPv6 to connect. If your network does not support IPv6, try any other internet network. In my experience, many mobile networks are supporting IPv6 technology, so try mobile network tethering/hotspots.

### How to test wether your network supports IPv6
- A reliable way to test would be to enter "ipv6.google.com" in the browser window. If google pops up, then IPv6 is working for you.

## How to customise the game using preferences
- In the main game menu, click preferences.
- Hover over each name to know more about them.

1. Sounds: When True, sounds are enabled.
2. Flip: When True, it shows the chess board from the perspective of the player who is playing, otherwise shows a constant board with white side at the bottom.
3. SlideShow: When True, it will show a slideshow of images on main menu.
4. Show moves: When True, it will show all legal move options for a selected piece during gameplay.
5. Undo: When True, it allows users to undo.
6. Show Clock: When True, it shows a clock in multiplayer chess mode when timer is disabled, clock displayes total time elapsed since start.

- You can also open res/preferences.txt and edit the file to your preferences

## What's Next
- I prefer to work on this app locally, I do not commit to github for every change I make. I only commit when a new version is available or when I update any readme, etc.
- For the next version (v3.3), I plan to release less of new features and focus on the GUI along with any bugfixes or performance improvements. Mainly because v3.2 has come with a lot of code refactoring, new features and changes to the backend, and GUI hasn't recieved much attention by me in this release.


CHANGELOG.md
# Changelog
This is the basic changelog for My-PyChess. All dates are in dd/mm/yyyy format.

## What's new in Version 3.2 (2/9/2020)
- Added a Back-Button, to go back to the previous menu. In older versions, the quit button was used for this purpose, but from now on a dedicated button to go back is there in the top-right corner. The quit button will now be used only to exit the app.
- Added a game timer to multiplayer mode, with a new menu to setup timer.
- Fixed bugs and made many additions and improvements to the client-server in chess online gamemode.
- Made optimisations to core chess module and gui module.
- Added a chess hotwto.
- Minor improvements to game sounds and textbox.
- Upgraded preference menu and made the loadgame interface more robust.
- Several other minor changes and improvements made.

## What was new in Version 3.1 (14/6/2020)
- Added basic sounds (beta).
- Added draw and resign option to online gameplay.
- Added a simple "About menu".
- Added compatibility for python v3.5 (server.py still needs python v3.6 to work)
- Made many changes to GUI throughout the app, including more prompt messages.
- Upgraded preferences.
- Fixed a small issue in move animations, added a new startgame animation.
- Fixed a bug where undo in singleplayer after game ended would give weird results.
- Fixed many bugs in online mode (server), made server robust to TCP packet loss among other things.

## What was new in Version 3.0 (23/5/2020)
- THIS WAS THE BIGGEST RELEASE OF MY-PYCHESS EVER RELEASED.
- The code came with MIT License instead of the GPL-3 which came with v2.2.
- The code was revamped, fixing minor bugs and MAJOR PERFORMANCE IMPROVEMENTS.
- On my PC, a chess match with v2.2 ran at 22 fps, while this upgrade allowed it to run at over 200 fps if fps was not constrained.
- Online play was added. This featured a ONLINE LOBBY and support upto 10 people to play chess(read more in ref/online.txt)
- Singleplayer saw BIG UPGRADES: A Good Menu, ability to play with a decent PYTHON CHESS ENGINE.
- Now, you can play against STOCKFISH CHESS ENGINE. My-PyChess will act as a GUI interface to give you this singleplayer mode. 
- Included with this release, a good menu for stockfish and stockfish install/configure menu to help with installing and configuring stockfish with My-PyChess.
- Code was made more easy to understand with code comments throughout the code.
- EnPassant, Screen flip and Undo moves option were added.
- UI for loadGame menu was revamped, UI for preference menu was simplified.
- Save/load games with a new way of storing gamedata.

## What was new in Version 2.3 (4/2/2020)
- This version never made it to github, I dropped it while working on it and started to focus on v3.0
- This had included a few minor GUI changes.

## What was new in Version 2.2 (22/1/2020)
- Minor performance upgrades.
- Further upgraded preferences.
- Upgraded loadgame menu (with delete games option).
- Minor changes made to the interface.
- While saving a game, it informed under what name the game was being saved.

## What was new in Version 2.1 (11/1/2020)
- The upgrade optimised the game a bit, fixing a bug.
- When you clicked a piece, it would show all the available spaces it could go.
- The upgrade added a new preference menu where you could customise some game features.
- The upgrade added background animations on the home menu.

## What was new in Version 2.0 (7/1/2020)
- Revamped the Home Menu so that it looks better.
- Revamped and code, optimised it.
- Fixed Bugs.
- Made some changes that ensure that it can work on all platforms.
- Added Save/load game features.
- Added highly basic SinglePlayer (random move generator)

# Changelog for older versions

Below, I will mention about the initial development of this app. At this state, the game was unstable - It lacked many features, had performance issues and lot, lot of bugs. Most of the versions given below have not made it to github.

## What was new in Version 1.5 (26/11/2019)
- added choice for pawn promotion
- added piece animations
- added some popup messages
- upgraded home menu
- fixed bugs

## What was new in Version 1.4 (9/11/2019)
- Added a home menu
- Added placeholders for future upgrades
- Basic compliance with pep8

## What was new in Version 1.3 (2/11/2019)
- coding optimisations
- Checkmate and stalemate report feature added
- fixed bugs

## What was new in Version 1.2 (24/10/2019)
- Improved move validation, stopping king from moving into attacked areas.
- fixed bugs

## What was new in Version 1.1 (21/10/2019)
- added castling, pawn-promotion
- This version fixed a few bugs, but introduced lot more bugs

## What was new in Version 1.0 (19/10/2019)
- Added Move validation for highly basic chess moves

## What was new in Version 0.3 (17/10/2019)
- Added a few more basic utility functions.
- Fixed a few bugs.
- Minor changes made to code.

## What was new in Version 0.2 (16/10/2019)
- Added GUI for pieces, implemented basic piece logic.

## What was new in Version 0.1 (15/10/2019)
- Added GUI for the chess board


loader.py
"""
This file is a part of My-PyChess application.
This file loads all the images and texts that are used.

Most of the scripts in this application import specific classes from this
module. Each class is a collection of resources for a particular script.
All font-related stuff is done in this file, the functions to put a number
on the screen and display date and time are also defined here

Level of development = STABLE
"""

import os.path
import pygame

# Initialize pygame.font module and load the font file.
pygame.font.init()
FONT = os.path.join("res", "Asimov.otf")

# Load different sizes of the font.
head = pygame.font.Font(FONT, 80)
large = pygame.font.Font(FONT, 50)
medium = pygame.font.Font(FONT, 38)
small = pygame.font.Font(FONT, 27)
vsmall = pygame.font.Font(FONT, 17)

# Define RGB color constants for use.
WHITE = (255, 255, 255)
GREY = (180, 180, 180)
BLACK = (0, 0, 0)
GREEN = (0, 255, 0)
RED = (200, 20, 20)

# Define a few constants that contain loaded texts of numbers and chararters.
NUM = [vsmall.render(str(i), True, WHITE) for i in range(10)]
LNUM = [small.render(str(i), True, WHITE) for i in range(10)]
SLASH = vsmall.render("/", True, WHITE)
COLON = vsmall.render(":", True, WHITE)

# This is a function that displays a number in a position on the screen.
# Very small sized text used.
def putNum(win, num, pos):
    for cnt, i in enumerate(list(str(num))):
        win.blit(NUM[int(i)], (pos[0] + (cnt * 8), pos[1]))


# This is a function that displays a number in a position on the screen.
# Small sized text used.
def putLargeNum(win, num, pos):
    for cnt, i in enumerate(list(str(num))):
        win.blit(LNUM[int(i)], (pos[0] + (cnt * 14), pos[1]))


# This is a function that displays the date and time in a position
# on the screen.
def putDT(win, DT, pos):
    var = DT.split()
    date = var[0].split("/")
    time = var[1].split(":")
    for cnt, num in enumerate(date):
        putNum(win, num, (pos[0] + 24 * cnt - 5, pos[1]))

    win.blit(SLASH, (pos[0] + 13, pos[1]))
    win.blit(SLASH, (pos[0] + 35, pos[1]))

    for cnt, num in enumerate(time):
        putNum(win, num, (pos[0] + 24 * cnt, pos[1] + 21))

    win.blit(COLON, (pos[0] + 20, pos[1] + 21))
    win.blit(COLON, (pos[0] + 43, pos[1] + 21))


# This splits a string at regular intervals of "index" characters
def splitstr(string, index=57):
    data = []
    while len(string) >= index:
        data.append(string[:index])
        string = string[index:]
    data.append(string)
    return data


# Defined important globals for loading background image sprites.
BGSPRITE = pygame.image.load(os.path.join("res", "img", "bgsprites.jpg"))
PSPRITE = pygame.image.load(os.path.join("res", "img", "piecesprite.png"))


class CHESS:
    PIECES = ({}, {})
    for i, ptype in enumerate(["k", "q", "b", "n", "r", "p"]):
        for side in range(2):
            PIECES[side][ptype] = PSPRITE.subsurface((i * 50, side * 50, 50, 50))

    CHECK = small.render("CHECK!", True, BLACK)
    STALEMATE = small.render("STALEMATE!", True, BLACK)
    CHECKMATE = small.render("CHECKMATE!", True, BLACK)
    LOST = small.render("LOST", True, BLACK)
    CHOOSE = small.render("CHOOSE:", True, BLACK)
    SAVE = small.render("Save Game", True, BLACK)
    UNDO = small.render("Undo", True, BLACK)

    MESSAGE = (
        small.render("Do you want to quit", True, WHITE),
        small.render("this game?", True, WHITE),
    )

    YES = small.render("YES", True, WHITE)
    NO = small.render("NO", True, WHITE)
    MSG = vsmall.render("Game will be saved with ID", True, WHITE)
    SAVE_ERR = vsmall.render("ERROR: SaveGame Limit Exeeded", True, WHITE)

    TURN = (
        small.render("Others turn", True, BLACK),
        small.render("Your turn", True, BLACK),
    )

    DRAW = small.render("Draw", True, BLACK)
    RESIGN = small.render("Resign", True, BLACK)


class LOADGAME:
    HEAD = large.render("Load Games", True, WHITE)
    LIST = medium.render("List of Games", True, WHITE)
    EMPTY = small.render("There are no saved games yet.....", True, WHITE)
    GAME = small.render("Game", True, WHITE)
    TYPHEAD = vsmall.render("Game Type:", True, WHITE)
    TYP = {
        "single": vsmall.render("SinglePlayer", True, WHITE),
        "mysingle": vsmall.render("SinglePlayer", True, WHITE),
        "multi": vsmall.render("MultiPlayer", True, WHITE),
    }
    DATE = vsmall.render("Date:", True, WHITE)
    TIME = vsmall.render("Time:", True, WHITE)

    DEL = pygame.image.load(os.path.join("res", "img", "delete.jpg"))
    LOAD = small.render("LOAD", True, WHITE)

    MESSAGE = (
        small.render("Are you sure that you", True, WHITE),
        small.render("want to delete game?", True, WHITE),
    )
    YES = small.render("YES", True, WHITE)
    NO = small.render("NO", True, WHITE)

    LEFT = medium.render("<", True, WHITE)
    RIGHT = medium.render(">", True, WHITE)
    PAGE = [medium.render("Page " + str(i), True, WHITE) for i in range(1, 5)]


class MAIN:
    HEADING = head.render("PyChess", True, WHITE)
    VERSION = vsmall.render("Version 3.1", True, WHITE)
    ICON = pygame.image.load(os.path.join("res", "img", "icon.gif"))
    BG = [BGSPRITE.subsurface((i * 500, 0, 500, 500)) for i in range(4)]

    SINGLE = medium.render("SinglePlayer", True, WHITE)
    MULTI = medium.render("MultiPlayer", True, WHITE)
    ONLINE = medium.render("Online", True, WHITE)
    LOAD = medium.render("Load Game", True, WHITE)
    ABOUT = medium.render("About", True, WHITE)
    PREF = medium.render("Preferences", True, WHITE)
    STOCK = small.render("Configure Stockfish", True, WHITE)

    SINGLE_H = medium.render("SinglePlayer", True, GREY)
    MULTI_H = medium.render("MultiPlayer", True, GREY)
    ONLINE_H = medium.render("Online", True, GREY)
    LOAD_H = medium.render("Load Game", True, GREY)
    ABOUT_H = medium.render("About", True, GREY)
    PREF_H = medium.render("Preferences", True, GREY)
    STOCK_H = small.render("Configure Stockfish", True, GREY)


class PREF:
    HEAD = large.render("Preferences", True, WHITE)

    SOUNDS = medium.render("Sounds", True, WHITE)
    FLIP = medium.render("Flip screen", True, WHITE)
    SLIDESHOW = medium.render("Slideshow", True, WHITE)
    MOVE = medium.render("Moves", True, WHITE)
    UNDO = medium.render("Allow undo", True, WHITE)

    COLON = medium.render(":", True, WHITE)

    TRUE = medium.render("True", True, WHITE)
    FALSE = medium.render("False", True, WHITE)

    SOUNDS_H = (
        vsmall.render("Play different sounds", True, WHITE),
        vsmall.render("and music", True, WHITE),
    )
    FLIP_H = (
        vsmall.render("This flips the screen", True, WHITE),
        vsmall.render("after each move", True, WHITE),
    )
    SLIDESHOW_H = (
        vsmall.render("This shows a slide of", True, WHITE),
        vsmall.render("backgrounds on screen", True, WHITE),
    )
    MOVE_H = (
        vsmall.render("This shows all the legal", True, WHITE),
        vsmall.render("moves of a selected piece", True, WHITE),
    )
    UNDO_H = (
        vsmall.render("This allowes undo if", True, WHITE),
        vsmall.render("set to be true", True, WHITE),
    )

    BSAVE = medium.render("Save", True, WHITE)
    TIP = vsmall.render(
        "TIP: Hover the mouse over the feature name to know", True, WHITE
    )
    TIP2 = vsmall.render("more about it.", True, WHITE)

    PROMPT = (
        vsmall.render("Are you sure you want to quit?", True, WHITE),
        vsmall.render("Any changes will not be saved.", True, WHITE),
    )

    YES = small.render("YES", True, WHITE)
    NO = small.render("NO", True, WHITE)


class ONLINE:
    TRYCONN = vsmall.render("Trying to connect to server....", True, WHITE)
    ERR = [
        vsmall.render("[ERR 1] Couldn't find the server...", True, WHITE),
        vsmall.render("[ERR 2] Server refused connection..", True, WHITE),
        vsmall.render("[ERR 3] Server is full (max = 10)..", True, WHITE),
        vsmall.render("[ERR 4] The server is Locked...", True, WHITE),
    ]
        
    EMPTY = small.render("No one's online, you are alone.", True, WHITE)

    LOBBY = large.render("Online Lobby", True, WHITE)
    LIST = medium.render("List of Players", True, WHITE)
    PLAYER = small.render("Player", True, WHITE)
    DOT = small.render(".", True, WHITE)

    ACTIVE = small.render("ACTIVE", True, GREEN)
    BUSY = small.render("BUSY", True, RED)
    REQ = small.render("Send Request", True, WHITE)
    YOUARE = medium.render("You Are", True, WHITE)

    REFRESH = pygame.image.load(os.path.join("res", "img", "refresh.png"))

    MSG1 = (
        vsmall.render("Please wait for the other player to", True, WHITE),
        vsmall.render("accept your request. Game will begin", True, WHITE),
        vsmall.render("shortly. You will play as white", True, WHITE),
    )
    MSG2 = (
        vsmall.render("Player", True, WHITE),
        vsmall.render("wants to play with you.", True, WHITE),
        vsmall.render("Accept to play. You will play as black", True, WHITE),
    )

    OPPQUIT = small.render("Your Opponent has left", True, WHITE)
    RESIGN = small.render("Your Opponent has resigned", True, WHITE)
    DRAWAGREED = small.render("Draw has been agreed", True, WHITE)

    DRAW = (
        vsmall.render("Sent a request to your opponent for", True, WHITE),
        vsmall.render("draw, wait for reply.", True, WHITE),
    )

    DRAW2 = (
        vsmall.render("Your opponent is requesting for a", True, WHITE),
        vsmall.render("draw, please reply.", True, WHITE),
    )

    NO = small.render("NO", True, WHITE)
    OK = small.render("OK", True, WHITE)


class ONLINEMENU:
    HEAD = large.render("Online", True, WHITE)
    with open(os.path.join("res", "texts", "online.txt")) as f:
        TEXT = [vsmall.render(i, True, WHITE) for i in f.read().splitlines()]

    with open(os.path.join("res", "texts", "onlinehowto.txt")) as f:
        TEXT2 = [vsmall.render(i, True, WHITE) for i in f.read().splitlines()]

    CLICK = vsmall.render("Click Here", True, WHITE)
    BACK = vsmall.render("Go Back", True, WHITE)
    CONNECT = small.render("Connect", True, WHITE)

class SINGLE:
    HEAD = large.render("Singleplayer", True, WHITE)
    _PARA1 = (
        "Play chess against a chess player algorithm implemented in",
        "python. This is called MiniMax algorithm and is used with",
        "alpha-beta optimisation technique. Currently, it is setup",
        "to play like a average (weak) player of chess.",
    )
    PARA1 = [vsmall.render(i, True, WHITE) for i in _PARA1]
    SELECT = pygame.image.load(os.path.join("res", "img", "select.png"))
    CHOOSE = small.render("Choose:", True, WHITE)
    START = small.render("Start Game", True, WHITE)
    OR = medium.render("OR", True, WHITE)
    _PARA2 = (
        "Play chess against the StockFish Chess Engine, the best",
        "chess engine in the world. This can play chess with a variety",
        "of difficulty settings, the hardest levels can easily defeat",
        "chess grandmasters.",
    )
    PARA2 = [vsmall.render(i, True, WHITE) for i in _PARA2]
    LEVEL = small.render("Level:", True, WHITE)

    BACK = vsmall.render("Go Back", True, WHITE)
    _CONFIG = (
        "It looks like you have not configured",
        "stockfish. To play, you have to do",
        "that.",
    )
    CONFIG = [vsmall.render(i, True, WHITE) for i in _CONFIG]
    OK = vsmall.render("Ok", True, WHITE)
    NOTNOW = vsmall.render("Not Now", True, WHITE)

class STOCKFISH:
    HEAD = large.render("Stockfish Engine", True, WHITE)
    CONFIG = small.render("Configure Stockfish", True, WHITE)
    with open(os.path.join("res", "texts", "stockfish.txt"), "r") as f:
        TEXT = [vsmall.render(i, True, WHITE) for i in f.read().splitlines()]

    with open(os.path.join("res", "texts", "configd.txt"), "r") as f:
        CONFIGURED = [vsmall.render(i, True, GREEN) for i in f.read().splitlines()]

    with open(os.path.join("res", "texts", "nonconfigd.txt"), "r") as f:
        NONCONFIGURED = [vsmall.render(i, True, RED) for i in f.read().splitlines()]

    CLICK = vsmall.render("Click Here", True, WHITE)
    BACK = vsmall.render("Go Back", True, WHITE)
    INSTALL = small.render("Install", True, WHITE)
    TEST = vsmall.render(
        "After all steps are complete, press button below.", True, WHITE
    )

    WIN_HEAD = small.render("Installation Guide for Windows", True, WHITE)
    LIN_HEAD = small.render("Installation Guide for Linux -", True, WHITE)
    MAC_HEAD = small.render("Installation Guide for Mac", True, WHITE)
    OTH_HEAD = small.render("Installation Guide for Other OS", True, WHITE)

    with open(os.path.join("res", "texts", "win.txt"), "r") as f:
        WIN_TEXT = [vsmall.render(i, True, WHITE) for i in f.read().splitlines()]

    with open(os.path.join("res", "texts", "linux.txt"), "r") as f:
        LIN_TEXT = [vsmall.render(i, True, WHITE) for i in f.read().splitlines()]

    with open(os.path.join("res", "texts", "linux2.txt"), "r") as f:
        LIN_TEXT2 = [vsmall.render(i, True, WHITE) for i in f.read().splitlines()]

    with open(os.path.join("res", "texts", "mac.txt"), "r") as f:
        MAC_TEXT = [vsmall.render(i, True, WHITE) for i in f.read().splitlines()]

    with open(os.path.join("res", "texts", "other.txt"), "r") as f:
        OTH_TEXT = [vsmall.render(i, True, WHITE) for i in f.read().splitlines()]

    for line in splitstr(os.path.abspath("res/stockfish/build/stockfish.exe")):
        WIN_TEXT.append(vsmall.render(line, True, WHITE))

    for line in splitstr(os.path.abspath("res/stockfish/build/stockfish")):
        LIN_TEXT2.append(vsmall.render(line, True, WHITE))
        OTH_TEXT.append(vsmall.render(line, True, WHITE))

    LOADING = head.render("Loading", True, WHITE)
    _SUCCESS = ("Setup successful, now you can go", "back and play chess.")
    _NOSUCCESS = (
        "Setup unsuccessful, try to re-",
        "configure. Follow instuctions",
        "carefully and try again.",
    )

    SUCCESS = [vsmall.render(i, True, GREEN) for i in _SUCCESS]
    NOSUCCESS = [vsmall.render(i, True, RED) for i in _NOSUCCESS]
    
    PROMPT = [
        small.render("Do you want to quit?", True, WHITE),
        vsmall.render("Stockfish is not configured yet.", True, WHITE)
    ]
    YES = small.render("Yes", True, WHITE)
    NO = small.render("No", True, WHITE)


class ABOUT:
    HEAD = large.render("About My-PyChess", True, WHITE)

    with open(os.path.join("res", "texts", "about.txt"), "r") as f:
        TEXT = [vsmall.render(i, True, WHITE) for i in f.read().splitlines()]

pygame.font.quit()
